core/src/main/java/org/apache/calcite/rel/rules/ExpandDisjuctionRule.java

```diff
@@ -0,0 +1,379 @@
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexTableInputRef;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ControlFlowException;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import com.google.common.collect.Lists;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.immutables.value.Value;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Rule to expand disjuction in condition of a {@link Filter} or {@link Join},
+ * It makes sense to make this optimization part of the predicate pushdown. For example:
+ *
+ * <blockquote><pre>{@code
+ * select t1.name from t1, t2
+ * where t1.id = t2.id
+ * and (
+ *  (t1.id > 20 and t2.height < 50)
+ *  or
+ *  (t1.weight < 200 and t2.sales > 100)
+ * )
+ * }</pre></blockquote>
+ *
+ * <p>we can expand to obtain the condition
+ *
+ * <blockquote><pre>{@code
+ * t1.id > 20 or t1.weight < 200
+ * t2.height < 50 or t2.sales > 100
+ * }</pre></blockquote>
+ *
+ * <p>new generated predicates are redundant, but they could be pushed down to
+ * scan operator of t1/t2 and reduce the cardinality.
+ *
+ * <p>This rule should only be applied once to avoid generate same redundant expression and
+ * it should be used before {@link CoreRules#FILTER_INTO_JOIN} and {@link CoreRules#JOIN_CONDITION_PUSH}.
+ *
+ * @see CoreRules#EXPAND_FILTER_DISJUCTION
+ * @see CoreRules#EXPAND_JOIN_DISJUCTION
+ */
+@Value.Enclosing
+public class ExpandDisjuctionRule
+    extends RelRule<ExpandDisjuctionRule.Config>
+    implements TransformationRule {
+
+  /**
+   * Creates a ExpandDisjuctionRule.
+   */
+  protected ExpandDisjuctionRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    config.matchHandler().accept(this, call);
+  }
+
+  /**
+   * Expand predicates for condition of a {@link Filter} or {@link Join}.
+   *
+   * @param condition   Condition to be expanded in Filter or Join
+   * @param relNode     The Filter or Join node
+   * @param fieldList   The field list of the Filter or Join inputs. Build the referenced columns
+   *                    in the predicate as RexInputRef and  find out which table the RexInputRef
+   *                    comes from via {@link RelMetadataQuery#getExpressionLineage}
+   * @param relBuilder  Builder
+   * @param mq          RelMetadataQuery, which is used to get the expression lineage
+   *
+   * @return Expanded predicates
+   */
+  private RexNode apply(
+      RexNode condition,
+      RelNode relNode,
+      List<RelDataTypeField> fieldList,
+      RelBuilder relBuilder,
+      RelMetadataQuery mq) {
+    HashMap<Integer, RexTableInputRef.RelTableRef> inputRefToTableRef = new HashMap<>();
+    ImmutableBitSet columnBits = RelOptUtil.InputFinder.bits(condition);
```

Comment:
Nit: you should group together (i.e., no newline) lines that make sense to be considered together (like a variable used in a loop coming right after it, or a variable used in the following statement), and split groups of lines that are conceptually separated, like "extract input ref and their mappings to tables" and "computing the new filter conditions".

I know it's not a deal breaker, but I find it quite hard to read as-is, it would be nice if you could do a pass over the code to improve readability.

Here, for instance, I'd put a newline between 112 and 113 (which can go with the `if` that follows as they are strictly related).

For formatting, you can generally take a look around at similar code to take inspiration.